Tasks/Npm/npmtask.ts

```diff
@@ -1,27 +1,190 @@
+import Q = require('q');
+import fs = require('fs');
 import path = require('path');
+import url = require('url');
 import tl = require('vsts-task-lib/task');
+import trm = require('vsts-task-lib/toolrunner');
+
+interface EnvironmentDictionary { [key: string]: string; }
 
 tl.setResourcePath(path.join( __dirname, 'task.json'));
 
-var npm = tl.createToolRunner(tl.which('npm', true));
+executeTask();
+
+async function executeTask() {
+
+    var cwd = tl.getPathInput('cwd', true, false);
+    tl.mkdirP(cwd);
+    tl.cd(cwd);
+
+    var command = tl.getInput('command', true);
+    if (command.indexOf(' ') >= 0) {
+        tl.setResult(tl.TaskResult.Failed, tl.loc("InvalidCommand"));
+    }
+
+    var npmRunner = tl.createToolRunner(tl.which('npm', true));
+    npmRunner.arg(command);
+    npmRunner.argString(tl.getInput('arguments', false));
+
+
+    if(shouldUseDeprecatedTask()) {
+
+        // deprecated version of task, which just runs the npm command with NO auth support.
+        try{
+            var code : number = await npmRunner.exec();
+            tl.setResult(code, tl.loc('NpmReturnCode', code));
+        } catch (err) {
+            tl.debug('taskRunner fail');
+            tl.setResult(tl.TaskResult.Failed, tl.loc('NpmFailed', err.message));
+        }
+    } else {
+
+        // new task version with auth support
+        var npmrcPath: string = path.join(cwd, '.npmrc');
+        var tempNpmrcPath : string = getTempNpmrcPath();
+        var debugLog: boolean = true;
+        if (tl.getVariable('system.debug') != undefined && tl.getVariable('system.debug').toLowerCase() === 'false') {
+            debugLog = false;
+        }
 
-var cwd = tl.getPathInput('cwd', true, false);
-tl.mkdirP(cwd);
-tl.cd(cwd);
+        try{
+            await runNpmAuthHelperAsync(getNpmAuthHelperRunner(npmrcPath, tempNpmrcPath, debugLog));
 
-var command = tl.getInput('command', true);
-if (command.indexOf(' ') >= 0) {
-	tl.setResult(tl.TaskResult.Failed, tl.loc("InvalidCommand"));
+            // set required environment variables for npm execution
+            var npmExecOptions: trm.IExecOptions = {};
+            npmExecOptions.env = process.env;
+            npmExecOptions.env['npm_config_userconfig'] = tempNpmrcPath;
+            if(debugLog) {
+                npmExecOptions.env['npm_config_loglevel'] =  'verbose';
+            }
+
+            await tryRunNpmConfigAsync(getNpmConfigRunner(debugLog), npmExecOptions);
+            var code : number =  await runNpmCommandAsync(npmRunner, npmExecOptions);
+            cleanUpTempNpmrcPath(tempNpmrcPath);
```

Comment:
setResults method below halts the execution without running the finally block. It's weird.